pkg/runtime/serializer/codec_factory.go

```diff
@@ -55,16 +55,75 @@ func newCodecFactory(scheme *runtime.Scheme, mf json.MetaFactory) CodecFactory {
 			Serializer:         jsonSerializer,
 			PrettySerializer:   jsonPrettySerializer,
 		},
-		{
-			AcceptContentTypes: []string{"application/yaml"},
-			ContentType:        "application/yaml",
-			FileExtensions:     []string{"yaml"},
-			Serializer:         yamlSerializer,
-		},
 	}
+	yamlSerializer := json.NewYAMLSerializer(mf, scheme, runtime.ObjectTyperToTyper(scheme))
+	serializers = append(serializers, serializerType{
+		AcceptContentTypes: []string{"application/yaml"},
+		ContentType:        "application/yaml",
+		FileExtensions:     []string{"yaml"},
+		Serializer:         yamlSerializer,
+	})
+
+	for _, fn := range serializerExtensions {
+		if serializer, ok := fn(scheme); ok {
+			serializers = append(serializers, serializer)
+		}
+	}
+	return serializers
+}
+
+// CodecFactory provides methods for retrieving codecs and serializers for specific
+// versions and content types.
+type CodecFactory struct {
+	scheme      *runtime.Scheme
+	serializers []serializerType
+	universal   runtime.Decoder
+	accepts     []string
+
+	legacySerializer runtime.Serializer
+}
+
+// NewCodecFactory provides methods for retrieving serializers for the supported wire formats
+// and conversion wrappers to define preferred internal and external versions. In the future,
+// as the internal version is used less, callers may instead use a defaulting serializer and
+// only convert objects which are shared internally (Status, common API machinery).
+// TODO: allow other codecs to be compiled in?
+// TODO: accept a scheme interface
+func NewCodecFactory(scheme *runtime.Scheme) CodecFactory {
+	serializers := newSerializersForScheme(scheme, json.DefaultMetaFactory)
+	return newCodecFactory(scheme, serializers)
+}
+
+// NewStreamingCodecFactory returns serializers that support the streaming.Serializer interface.
+func NewStreamingCodecFactory(scheme *runtime.Scheme) CodecFactory {
+	return newStreamingCodecFactory(scheme, json.DefaultMetaFactory)
+}
+
+// newStreamingCodecFactory handles providing streaming codecs
+func newStreamingCodecFactory(scheme *runtime.Scheme, mf json.MetaFactory) CodecFactory {
+	serializers := newSerializersForScheme(scheme, mf)
+	streamers := []serializerType{}
+	for i := range serializers {
+		if serializers[i].ImplicitSerializer != nil {
+			serializers[i].Serializer = serializers[i].ImplicitSerializer
+		}
+		if s, ok := serializers[i].Serializer.(streaming.Framer); ok {
```

Comment:
That's in the watch branch (not a PR yet), I will add the comment here.

On Thu, Mar 10, 2016 at 7:43 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/runtime/serializer/codec_factory.go
> <https://github.com/kubernetes/kubernetes/pull/20377#discussion_r55673068>
> :
>
> > +}
> > +
> > +// NewStreamingCodecFactory returns serializers that support the streaming.Serializer interface.
> > +func NewStreamingCodecFactory(scheme *runtime.Scheme) CodecFactory {
> > +	return newStreamingCodecFactory(scheme, json.DefaultMetaFactory)
> > +}
> > +
> > +// newStreamingCodecFactory handles providing streaming codecs
> > +func newStreamingCodecFactory(scheme *runtime.Scheme, mf json.MetaFactory) CodecFactory {
> > +	serializers := newSerializersForScheme(scheme, mf)
> > +	streamers := []serializerType{}
> > +	for i := range serializers {
> > +		if serializers[i].ImplicitSerializer != nil {
> > +			serializers[i].Serializer = serializers[i].ImplicitSerializer
> > +		}
> > +		if s, ok := serializers[i].Serializer.(streaming.Framer); ok {
>
> IIUC, nothing is implementing Framer interface for now. That means
> streamers will always be empty.
> Can you please add a TODO in json/yaml/proto serializer to make the
> implements streaming.Framer?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20377/files#r55673068>.
>